packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts

```diff
@@ -0,0 +1,372 @@
+      return null;
+    }
+
+    const directive = this.getDirectiveSymbolForAccessExpression(node.left);
+    const type = this.typeChecker.getTypeOfSymbolAtLocation(tsSymbol, node);
+    return {
+      kind: SymbolKind.Input,
+      bindings: directive === null ? [] : [{
+        kind: SymbolKind.Binding,
+        tsSymbol,
+        tsType: type,
+        directive,
+        shimLocation: {shimPath: this.shimPath, positionInShimFile},
+      }],
+    };
+  }
+
+  private getDirectiveSymbolForAccessExpression(node: ts.ElementAccessExpression|
+                                                ts.PropertyAccessExpression): DirectiveSymbol|null {
+    // In either case, `_t1["index"]` or `_t1.index`, `node.expression` is _t1.
+    // The retrieved symbol for _t1 will be the variable declaration.
+    const declaration = this.typeChecker.getSymbolAtLocation(node.expression)?.declarations[0];
+    if (declaration === undefined || !ts.isVariableDeclaration(declaration)) {
+      return null;
+    }
+
+    const symbol = this.getSymbolOfVariableDeclaration(declaration);
+    if (symbol === null || symbol.tsSymbol === null) {
+      return null;
+    }
+
+    return {
+      kind: SymbolKind.Directive,
+      tsSymbol: symbol.tsSymbol,
+      tsType: symbol.tsType,
+      shimLocation: symbol.shimLocation,
+    };
+  }
+
+  private getSymbolOfVariable(variable: TmplAstVariable): VariableSymbol|null {
+    const node = findFirstMatchingNode(
+        this.typeCheckBlock, {withSpan: variable.sourceSpan, filter: ts.isVariableDeclaration});
+    if (node === null) {
+      return null;
+    }
+
+    const expressionSymbol = this.getSymbolOfVariableDeclaration(node);
+    if (expressionSymbol === null) {
+      return null;
+    }
+
+    return {...expressionSymbol, kind: SymbolKind.Variable, declaration: variable};
+  }
+
+  private getSymbolOfReference(ref: TmplAstReference): ReferenceSymbol|null {
+    const target = this.templateData.boundTarget.getReferenceTarget(ref);
+    // Find the node for the reference declaration, i.e. `var _t2 = _t1;`
+    let node = findFirstMatchingNode(
+        this.typeCheckBlock, {withSpan: ref.sourceSpan, filter: ts.isVariableDeclaration});
+    if (node === null || target === null || node.initializer === undefined) {
+      return null;
+    }
+    // References that target directives or elements will be assigned to the TCB var for the
+    // directive/element instance.
+    // var _t1 = document.createElement('div');
+    // var _t2 = _t1;
+    // Getting the symbol for the initializer will give us the variable declaration node for
+    // the div creation. That's the symbol we want to return for the reference.
+    if (!(target instanceof TmplAstTemplate)) {
+      const initializerSymbol = this.typeChecker.getSymbolAtLocation(node.initializer);
+      if (initializerSymbol === undefined || initializerSymbol.declarations.length === 0 ||
+          ts.isVariableDeclaration(initializerSymbol.declarations[0])) {
```

Comment:
This looks a bit broken currently:

```suggestion
          !ts.isVariableDeclaration(initializerSymbol.declarations[0])) {
```

Should also allow dropping the cast below.